### PR TITLE
ProblyWrong: unified-evidential training first prototype and notebooks showcasing the use of it

### PR DESCRIPTION
**Pull request by Abdoul1710**
Implementation of a first prototype for a unified evidential training function. And notebooks covering every paper of [Issue #25](https://github.com/Cortys/SEP-Probly-WS2526/issues/25#issue-3586873531)
while using our new unified-evidential-train function

## Issue
The Issues of every notebooks are:
Unified Evidential Training Function (based on Sensory et al. 2018): https://github.com/Cortys/SEP-Probly-WS2526/issues/74

Prior networks (Malinin & Gales, 2018): https://github.com/Cortys/SEP-Probly-WS2526/issues/186
Regression Prior Network (Malinin et al., 2020): https://github.com/Cortys/SEP-Probly-WS2526/issues/189
Natural Posterior Network (Charpentier et al., 2022): https://github.com/Cortys/SEP-Probly-WS2526/issues/188
Information Robust Dirichlet Networks (Tsiligkaridis, 2019): https://github.com/Cortys/SEP-Probly-WS2526/issues/187
Posterior Networks (Charpentier et al., 2020): https://github.com/Cortys/SEP-Probly-WS2526/issues/185
Deep Evidential Regression (Amini et al., 2020): https://github.com/Cortys/SEP-Probly-WS2526/issues/184

## Public API Changes

-   [ ] No Public API changes
-   [x] Yes, Public API changes (Details below)

created new package probly.layers.evidential
created new losses package and moved all existing loss functions from probly.train.evidential.torch into probly.losses.evidential.torch
created new package probly.models

## How Has This Been Tested?

tested the unified-evidential-train function by using it in the notebooks and succesfully gaining losses which could be transformed into uncertainties

## Checklist

-   [x] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to [`CHANGELOG.md`](https://github.com/pwhofman/probly/blob/main/CHANGELOG.md) (if relevant for users).
-   [x] The code follows the project's [style guidelines](https://github.com/pwhofman/probly/blob/main/.github/CONTRIBUTING.md) and passes minimal code style checks.
-   [x] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

**Comment by Abdoul1710** [comment]
es failt ein test wegen einem notebook was gar nicht existiert? ein dropout_tutorial, was in notebook/examples liegen soll aber da ist keins und wir haben keins erstellt dafür...

File: src/probly/layers/evidential/torch.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+        if self.gamma_bias is not None:
+            fan_in, _ = init._calculate_fan_in_and_fan_out(self.gamma)  # noqa: SLF001
+            bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
+            init.uniform_(self.gamma_bias, -bound, bound)
+            init.uniform_(self.nu_bias, -bound, bound)
+            init.uniform_(self.alpha_bias, -bound, bound)
+            init.uniform_(self.beta_bias, -bound, bound)
+
+
+class SimpleCNN(nn.Module):
```
Comment (Cortys) [inline src/probly/layers/evidential/torch.py]: `probly.layers` should only contain reusable layers, not dataset-specific model implementations. Please remove this; demo models should be part of notebooks, or, if used for testing, moved into a test fixture.

File: src/probly/train/evidential/common.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+from torch import nn
+
+from probly.train.evidential.torch import der_loss, rpn_ng_kl, rpn_prior, train_pn
+
+if TYPE_CHECKING:
+    from torch.utils.data import DataLoader
+
+
+def unified_evidential_train_class(  # noqa: C901, PLR0912
+    mode: str,
```
Comment (Cortys) [inline src/probly/train/evidential/common.py]: Use `Literal["PostNet", "NatPostNet", "EDL", ...]` here

File: src/probly/train/evidential/torch.py [hunk 3]
```diff
@@ -1,169 +1,919 @@
...
+        Returns:
+            torch.Tensor, scalar loss value
         """
-        loss = (torch.abs(targets - inputs["gamma"]) * (2 * inputs["nu"] + inputs["alpha"])).mean()
-        return loss
+        num_classes = alpha_pred.shape[1]
+        device = alpha_pred.device
+
+        # In-distribution loss: KL(target_sharp || predicted)
+        alpha_target_in = self._make_in_domain_target_alpha(y, num_classes)
```
[the remark this answers is not in the document]
Comment (Abdoul1710) [reply src/probly/train/evidential/torch.py]: can't fine this piece of code anymore (maybe removed this part in previous commit)

File: src/probly/train/evidential/common.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            loss.backward()  # backpropagation
+            optimizer.step()  # updates model-parameters
+
+            total_loss += loss.item()  # add-up the loss of this epoch ontop of our total loss
+
+        avg_loss = total_loss / len(dataloader)  # calculate average loss per epoch across all batches
+        print(f"Epoch [{epoch + 1}/{epochs}] - Loss: {avg_loss:.4f}")  # noqa: T201
+
+
+def unified_evidential_train_reg(
```
[the remark this answers is not in the document]
Comment (Abdoul1710) [reply src/probly/train/evidential/common.py]: removed function. merged it with train_unified_evidential_train

File: src/probly/train/evidential/common.py
[the remark this answers is not in the document]
Comment (Abdoul1710) [reply src/probly/train/evidential/common.py]: seperated torch and common code and implemented use of lazy-dispatch

File: src/probly/train/evidential/torch.py [hunk 5]
```diff
@@ -1,169 +1,144 @@
...
+@_uet_dispatch.register(torch.nn.Module)
+def _torch_uet(  # noqa: C901, PLR0912, PLR0915
+    model: nn.Module,
+    *,
+    mode: Literal["PostNet", "NatPostNet", "EDL", "PrNet", "IRD", "DER", "RPN"],
+    dataloader: DataLoader,
+    loss_fn: torch.Tensor = None,
+    oodloader: DataLoader = None,
+    flow: torch.Tensor = None,
+    class_count: torch.Tensor = None,
```
Comment (Cortys) [inline src/probly/train/evidential/torch.py]: Incorrect type hints. Add `| None` to explicitly support optionals.

Comment (Abdoul1710) [reply src/probly/train/evidential/torch.py]: changed loss_fn to Callable[..., torch.Tensor] and corrected defaults

File: src/probly/train/evidential/common.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        mode=mode,
+        dataloader=dataloader,
+        loss_fn=loss_fn,
+        oodloader=oodloader,
+        flow=flow,
+        class_count=class_count,
+        epochs=epochs,
+        lr=lr,
+        device=device,
+    )
```
[the remark this answers is not in the document]
Comment (Abdoul1710) [reply src/probly/train/evidential/common.py]: moved

File: src/probly/layers/evidential/torch.py [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+            nn.ReLU(),
+            nn.Linear(256, num_classes),
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Return Dirichlet concentration parameters (alpha)(x) > 0."""
+        x = self.features(x)
+        logits = self.classifier(x)
+        alpha = F.softplus(logits) + 1e-3
+        return alpha
```
[the remark this answers is not in the document]
Comment (Abdoul1710) [reply src/probly/layers/evidential/torch.py]: removed

File: src/probly/losses/evidential/torch.py [hunk 8]
```diff
@@ -0,0 +1,666 @@
...
+
+    err = (y - p) ** 2
+    var = p * (1 - p) / (strengths[:, None] + 1)
+
+    loss = torch.mean(torch.sum(err + var, dim=1))
+
+    return loss
+
+
+def evidential_kl_divergence(inputs: Tensor, targets: Tensor) -> Tensor:
```
Comment (Cortys) [inline src/probly/losses/evidential/torch.py]: See previous comment regarding alpha vs evidence inputs.

File: src/probly/models/evidential/torch.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Encode inputs into latent representations.
+
+        Args:
+            x: Input tensor of shape (B, ...) where B is batch size.
+
+        Returns:
+            Latent tensor z of shape (B, latent_dim).
+        """
+        return self.encoder(x)
```
Comment (Cortys) [inline src/probly/models/evidential/torch.py]: This is not a posterior network.

File: src/probly/layers/evidential/torch.py [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+    """Head mapping latent features to Dirichlet concentration parameters."""
+
+    def __init__(self, latent_dim: int, num_classes: int) -> None:
+        """Initialize the Dirichlet classification head."""
+        super().__init__()
+
+        self.net = nn.Sequential(
+            nn.Linear(latent_dim, 256),
+            nn.ReLU(),
+            nn.Linear(256, num_classes),
```
Comment (Cortys) [inline src/probly/layers/evidential/torch.py]: hardcoded layer widths should be avoided

File: src/probly/models/evidential/torch.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        Args:
+            encoder: Encoder module mapping inputs to latent space.
+            head: Dirichlet head module mapping latent features to alpha parameters.
+            num_classes: Number of output classes.
+            latent_dim: Latent dimension for encoder (default: 128).
+        """
+        super().__init__()
+
+        if head is None:
+            self.head = t.IRDHead(latent_dim=latent_dim, num_classes=num_classes)
```
Comment (Cortys) [inline src/probly/models/evidential/torch.py]: Why did you default to the IRDHead here?

File: src/probly/models/evidential/torch.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+                num_classes=num_classes,
+            )
+
+        self.encoder = encoder
+        self.head = head
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Compute Dirichlet parameters for input samples."""
+        z = self.encoder(x)
+        return self.head(z)
```
[the remark this answers is not in the document]
Comment (Abdoul1710) [reply src/probly/models/evidential/torch.py]: renamed to PrNerModel to make clear thats its needed as default for PrNet-Paper
